In [146]:
#@unpublished{SARC,
#  authors={Mikhail Khodak and Nikunj Saunshi and Kiran Vodrahalli},
#  title={A Large Self-Annotated Corpus for Sarcasm},
#  url={https://arxiv.org/abs/1704.05579},
#  year=2017
#}
#https://www.kaggle.com/danofer/sarcasm#train-balanced-sarcasm.csv
import os
import string
import numpy as np
import pandas as pd
import nltk
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to C:\Users\Luba
[nltk_data]     Tovbin\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to C:\Users\Luba
[nltk_data]     Tovbin\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [203]:
# df - data frame
df = pd.read_csv('C:/Users/Luba Tovbin/Desktop/CMPE-257/Team_Proj_Sarcasm/sarcasm/train-balanced-sarcasm.csv')
# dropping empty comment entries
df.dropna(subset=['comment'], inplace=True)

print('Type:')
print('Labels array ',type(df.label))
print('Comments array ',type(df.comment))
print('one comment line ', type(df.comment[0]))
print('Shape:')
print('Labels array ',df.label.shape)
print('Comments array ',df.comment.shape)
print('Two first entries:')
print (df.label[0], df.comment[0])
print (df.label[1], df.comment[1])

Type:
Labels array  <class 'pandas.core.series.Series'>
Comments array  <class 'pandas.core.series.Series'>
one comment line  <class 'str'>
Shape:
Labels array  (1010773,)
Comments array  (1010773,)
Two first entries:
0 NC and NH.
0 You do know west teams play against west teams more than east teams right?


In [148]:
from sklearn.feature_extraction.text import TfidfVectorizer
# create the transform
tf_idf_vectorizer = TfidfVectorizer()
# tokenize and build vocab
T = tf_idf_vectorizer.fit(df.comment)
print(len(T.vocabulary_))

167435


In [208]:
# preprocess with nltk
def my_tokenizer(corpus):
    corpus_tokenized = []
    tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+')
    sbs = nltk.stem.SnowballStemmer('english', ignore_stopwords=False)
    for comment in corpus:
        words = tokenizer.tokenize(comment)
        cmnt_t = []
        for token in words:
            cmnt_t.append(sbs.stem(token))
            # make a string to be compatible with TfidfVectorizer
            c = ' '.join(cmnt_t)
        corpus_tokenized.append(c)
    return corpus_tokenized

In [223]:
df2 = my_tokenizer(df.comment)

In [224]:
for i in range(10):
    print(df.comment[i])
    print(df2[i], '\n')

NC and NH.
nc and nh 

You do know west teams play against west teams more than east teams right?
you do know west team play against west team more than east team right 

They were underdogs earlier today, but since Gronk's announcement this afternoon, the Vegas line has moved to patriots -1
they were underdog earlier today but sinc gronk s announc this afternoon the vega line has move to patriot 1 

This meme isn't funny none of the "new york nigga" ones are.
this meme isn t funni none of the new york nigga one are 

I could use one of those tools.
i could use one of those tool 

I don't pay attention to her, but as long as she's legal I wouldn't kick her out of bed (before she took a load)
i don t pay attent to her but as long as she s legal i wouldn t kick her out of bed befor she took a load 

Trick or treating in general is just weird...
trick or treat in general is just weird 

Blade Mastery+Masamune or GTFO!
blade masteri masamun or gtfo 

You don't have to, you have a good buil

In [225]:
print(df2[0:5], '\n')
# convert to pandas Series type type to be compatible with TfidfVectorizer
df3 = pd.Series((v for v in df2))
print(type(df3), '\n')
print(type(df3[0]))
print(df3[0:5])

['nc and nh', 'you do know west team play against west team more than east team right', 'they were underdog earlier today but sinc gronk s announc this afternoon the vega line has move to patriot 1', 'this meme isn t funni none of the new york nigga one are', 'i could use one of those tool'] 

<class 'pandas.core.series.Series'> 

<class 'str'>
0                                            nc and nh
1    you do know west team play against west team m...
2    they were underdog earlier today but sinc gron...
3    this meme isn t funni none of the new york nig...
4                        i could use one of those tool
dtype: object


In [226]:
tf_idf_vectorizer2 = TfidfVectorizer()
T2 = tf_idf_vectorizer2.fit(df3)
print(len(T2.vocabulary_))

131022


167435 features from before now reduced to 131022 features, that is about 22%

In [ ]:
from sklearn.model_selection import train_test_split
# divide into train and test sets
X_train, X_test, y_train, y_test = train_test_split(df.comment, df.label, train_size=0.33)

In [237]:
# Vetorize the training data set
X_train = tf_idf_vectorizer2.transform(X_train)
print(X_train.shape)
print(X_train[0:3])

(333555, 131022)
  (0, 114779)	0.1793130137737192
  (0, 58644)	0.9837920730984704
  (1, 114779)	0.2512595263266975
  (1, 80766)	0.9679197541274193
  (2, 127537)	0.18495877911554362
  (2, 126592)	0.1436433977478119
  (2, 121207)	0.14403305142069298
  (2, 114779)	0.13743171050627323
  (2, 114719)	0.08868106246619416
  (2, 95725)	0.36071037297517844
  (2, 85499)	0.16344319143927794
  (2, 83739)	0.09397292171976926
  (2, 80846)	0.1613647315882898
  (2, 71548)	0.23700624386247882
  (2, 68870)	0.11984800205110516
  (2, 60384)	0.09157056666960955
  (2, 58715)	0.10027337947697297
  (2, 49674)	0.1432978982615192
  (2, 44760)	0.22451918581003313
  (2, 40784)	0.1790493202911679
  (2, 35441)	0.1669070294508233
  (2, 33059)	0.25360823639242763
  (2, 29534)	0.2522865539551292
  (2, 26464)	0.2214490172017322
  (2, 22865)	0.1235961067028143
  (2, 16301)	0.4157599963225752
  (2, 12882)	0.29800211514232305
  (2, 10707)	0.17940520752579436


In [108]:
# Vetorize the test data set
X_test = tf_idf_vectorizer.transform(X_test)
print(X_test.shape)
print(X_test[0:3])

(677218, 167435)
  (0, 147353)	0.42468250459405066
  (0, 135515)	0.6237320637071776
  (0, 121397)	0.5117743421335358
  (0, 103526)	0.3065539156995408
  (0, 68465)	0.27334008578760505
  (1, 136405)	0.3003210486437824
  (1, 127586)	0.30946562001365147
  (1, 117995)	0.4490575503323284
  (1, 115370)	0.2656883076897965
  (1, 114176)	0.2745484447420461
  (1, 106800)	0.1471396221493557
  (1, 74902)	0.12746855084064718
  (1, 60734)	0.20408522519434966
  (1, 53175)	0.3201515492324172
  (1, 34530)	0.3698623578316731
  (1, 29842)	0.2618033452953899
  (1, 18307)	0.28112851168170877
  (2, 70503)	0.33850766790092285
  (2, 7708)	0.9409636330763684


In [238]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import BaggingClassifier
# bagging with Logistic Regressoin
# oob Out-Of-Bag
bag_log_reg = BaggingClassifier(
      LogisticRegression(), n_estimators=50,
    max_samples=50000, bootstrap=True, n_jobs=-1, oob_score=True)

bag_log_reg.fit(X_train,y_train)
print('oob_score is ', bag_log_reg.oob_score_)

oob_score is  0.6585060934478572


In [239]:
from sklearn.ensemble import RandomForestClassifier
rnd_frst = RandomForestClassifier(
      n_estimators=100, max_leaf_nodes=1000, n_jobs=-1, oob_score=True)
rnd_frst.fit(X_train,y_train)
print('oob_score is ', bag_log_reg.oob_score_)

KeyboardInterrupt: 